rfd/0151-database-permission-management.md

```diff
@@ -0,0 +1,355 @@
+---
+authors: Krzysztof Skrzętnicki (krzysztof.skrzetnicki@goteleport.com)
+state: draft
+---
+
+# RFD 151 - Database Permission Management
+
+## Required Approvals
+
+- Engineering: @r0mant && @smallinsky
+- Product: @xinding33 || @klizhentas
+- Security: @reedloden || @jentfoo
+
+## What
+
+Extends [automated db user provisioning](0113-automatic-database-users.md) with
+permission management capabilities.
+
+## Why
+
+Managing database-level permissions is a natural extension of current Teleport
+RBAC capabilities. The described model should also integrate seamlessly with
+TAG, providing a future-proof solution.
+
+Database administrators will be able to use Teleport for both user and
+permission management.
+
+## Details
+
+### UX
+
+A set of global "database object import rules" resources are defined for the
+Teleport cluster. At certain points in time, the database schema is read and
+passed through the import rules. The individual import rules may apply custom
+labels. An database object resource will only be created if there is at least
+one import rule that matched it.
+
+Only object attributes (standard ones like `protocol` or custom from the
+`attributes` field), which are sourced from the object spec, are subject to
+matching against the import rules. Any labels present on an object, either from
+an import rule or another source, are not subject to be matched by the import
+rule. The permissions matching is different in this regard.
+
+Example: the `sales-prod` import rule, which imports all tables from `sales`
+Postgres database in prod:
+
+```yaml
+kind: db_object_import_rule
+metadata:
+  name: sales-prod
+spec:
+  db_labels:
+    env: prod
+  mappings:
+    - object_match:
+        - database: sales
+          object_kind: table
+          protocol: postgres
+      add_labels:
+        env: prod
+        product: sales
+  priority: 10
+version: v1
+```
+
+Example database object, with applied labels:
+
+```yaml
+kind: db_object
+metadata:
+  labels:
+    env: prod
+    product: sales
+  name: sales_main
+spec:
+  attributes:
+    attr1: custom attr1 value
+    attr2: custom attr2 value
+  database: sales
+  name: sales_main
+  object_kind: table
```

Comment:
Sorry if I missed this somewhere, besides `table`, any other object_kind will be supported in the first version of implementation?

I assume the `attributes` is for future-proofing? Or are any custom `attributes` getting added to postgres `table`? If not, it would be great to provide a more practical example on what attributes could have, if possible

Reply:
> Sorry if I missed this somewhere, besides table, any other object_kind will be supported in the first version of implementation?

Ah, I didn't consider doing this for the first version -- I wanted to keep the implementation as simple as possible for MVP -- but eventually I think we can match the full coverage from this [table](https://www.postgresql.org/docs/current/ddl-priv.html#PRIVILEGES-SUMMARY-TABLE):

```
Object Type	All Privileges	Default PUBLIC Privileges	psql Command
DATABASE	CTc	Tc	\l
DOMAIN	U	U	\dD+
FUNCTION or PROCEDURE	X	X	\df+
FOREIGN DATA WRAPPER	U	none	\dew+
FOREIGN SERVER	U	none	\des+
LANGUAGE	U	U	\dL+
LARGE OBJECT	rw	none	\dl+
PARAMETER	sA	none	\dconfig+
SCHEMA	UC	none	\dn+
SEQUENCE	rwU	none	\dp
TABLE (and table-like objects)	arwdDxt	none	\dp
Table column	arwx	none	\dp
TABLESPACE	C	none	\db+
TYPE	U	U	\dT+
```

Some of these are rather niche, so perhaps can be skipped. Others will require further design (e.g. columns).

> I assume the attributes is for future-proofing? Or are any custom attributes getting added to postgres table? If not, it would be great to provide a more practical example on what attributes could have, if possible

The idea behind `attributes` field is to provide a way for the database engine discovering the object a way to express custom object metadata not captured by any existing fields. The alternative is to add fields one by one as needed by this engine or another, but I feel like this is too rigid. I'll try to provide some examples, though, to make this more concrete.

Reply:
One of reasons I am asking is, for postgres tables imported tables from a database, only `name` and `schema` will have different values?

Reply:
> One of reasons I am asking is, for postgres tables imported tables from a database, only `name` and `schema` will have different values?

Sure, that might be the case. Although you may want to enrich the information further: for example, add information about owner or other metadata, at which point the objects start to differ more.

Reply:
I've removed custom attributes; we can always add those if needed, but looks like the potential confusion outweighs the benefits.